realm/services/google_oauth/incoming_webhooks/oauth/source.js

```diff
@@ -35,56 +31,70 @@ exports = async function (payload, response) {
+
+    response.setStatusCode(302);
+    response.setHeader('Location', oauthURL.href);
   } else {
-    const oauthCode = payload.query.code;
-    if (typeof oauthCode === 'undefined') {
-      // No code provided, so let's request one from Google:
-      const oauthURL = new URL(GOOGLE_OAUTH_ENDPOINT);
-      oauthURL.search = querystring.stringify({
-        'client_id': CLIENT_ID,
-        'redirect_uri': OAUTH2_CALLBACK,
-        'response_type': 'code',
-        'scope': SCOPES.join(' '),
-        'access_type': "offline",
-      });
-
-      response.setStatusCode(302);
-      response.setHeader('Location', oauthURL.href);
-    } else {
-      // We have a code, so we've redirected successfully from Google's consent page.
-      // Let's post to Google, requesting an access:
-      let res = await context.http.post({
-        url: GOOGLE_TOKEN_ENDPOINT,
-        body: {
-          client_id: CLIENT_ID,
-          client_secret: CLIENT_SECRET,
-          code: oauthCode,
-          grant_type: 'authorization_code',
-          redirect_uri: OAUTH2_CALLBACK,
-        },
-        encodeBodyAsJSON: true,
-      });
-
-      // TODO: Need to handle errors here!
-
-      let tokens = JSON.parse(res.body.text());
-      tokens.updated = new Date();
-      tokens.expires_at = new Date();
-      tokens.expires_at.setTime(Date.now() + (tokens.expires_in * 1000));
-
-      await context.services.get("mongodb-atlas").db("auth").collection("auth_tokens").findOneAndReplace(
-        {
-          _id: "youtube"
-        },
-        tokens,
-        {
-          upsert: true,
-        },
-      );
+    // We have a code, so we've redirected successfully from Google's consent page.
+    // Let's post to Google, requesting an access:
+    let res = await context.http.post({
+      url: GOOGLE_TOKEN_ENDPOINT,
+      body: {
+        client_id: CLIENT_ID,
+        client_secret: CLIENT_SECRET,
+        code: oauthCode,
+        grant_type: 'authorization_code',
+        redirect_uri: OAUTH2_CALLBACK,
+      },
+      encodeBodyAsJSON: true,
+    });
+
+    let tokens = JSON.parse(res.body.text());
+
+    if (typeof tokens.expires_in === "undefined") {
+      throw new Error("Error response from Google: " + JSON.stringify(tokens))
+    }
 
+
+    if (typeof tokens.refresh_token === "undefined") {
       return {
-        "message": "ok",
-      }
+        "message": `You appear to have already linked to Google. You may need to revoke your OAuth token (${tokens.access_token}) and delete your auth token document. https://developers.google.com/identity/protocols/oauth2/web-server#tokenrevoke`
+      };
+    }
+
+    tokens._id = "youtube";
+    tokens.updated = new Date();
+    tokens.expires_at = new Date();
+    tokens.expires_at.setTime(Date.now() + (tokens.expires_in * 1000));
+
+    const tokens_collection = context.services.get("mongodb-atlas").db("auth").collection("auth_tokens");
+
+    // Upserts don't appear to work in Realm, so here's a hack insted:
```

Comment:
In case it's helpful, I was able to get upsert working with update() in get_youtube_video_stats

Reply:
I need to write a test script. It may be my misunderstanding.